src/util/test_util/mock_vm.rs

```diff
@@ -0,0 +1,597 @@
+                std::mem::size_of::<usize>()
+            })),
+            get_object_align_offset_when_copied: MockMethod::new_fixed(Box::new(|_| 0)),
+            get_type_descriptor: MockMethod::new_unimplemented(),
+            get_object_reference_when_copied_to: MockMethod::new_unimplemented(),
+            ref_to_object_start: MockMethod::new_fixed(Box::new(|object| {
+                object.to_raw_address().sub(DEFAULT_OBJECT_REF_OFFSET)
+            })),
+            ref_to_header: MockMethod::new_fixed(Box::new(|object| object.to_raw_address())),
+            ref_to_address: MockMethod::new_fixed(Box::new(|object| {
+                object.to_raw_address().sub(DEFAULT_OBJECT_REF_OFFSET)
+            })),
+            address_to_ref: MockMethod::new_fixed(Box::new(|addr| {
+                ObjectReference::from_raw_address(addr.add(DEFAULT_OBJECT_REF_OFFSET))
+            })),
+            dump_object: MockMethod::new_unimplemented(),
+
+            weakref_clear_referent: MockMethod::new_unimplemented(),
+            weakref_get_referent: MockMethod::new_unimplemented(),
+            weakref_set_referent: MockMethod::new_unimplemented(),
+            weakref_is_referent_cleared: MockMethod::new_fixed(Box::new(|r| r.is_null())),
+            weakref_enqueue_references: MockMethod::new_unimplemented(),
+
+            support_edge_enqueuing: MockMethod::new_fixed(Box::new(|_| true)),
+            scan_object: MockMethod::new_unimplemented(),
+            scan_object_and_trace_edges: MockMethod::new_unimplemented(),
+            scan_roots_in_mutator_thread: Box::new(MockMethod::<
+                (
+                    VMWorkerThread,
+                    &'static mut Mutator<MockVM>,
+                    ProcessEdgesWorkRootsWorkFactory<
+                        MockVM,
+                        SFTProcessEdges<MockVM>,
+                        SFTProcessEdges<MockVM>,
+                    >,
+                ),
+                (),
+            >::new_unimplemented()),
+            scan_vm_specific_roots: Box::new(MockMethod::<
+                (
+                    VMWorkerThread,
+                    ProcessEdgesWorkRootsWorkFactory<
+                        MockVM,
+                        SFTProcessEdges<MockVM>,
+                        SFTProcessEdges<MockVM>,
+                    >,
+                ),
+                (),
+            >::new_unimplemented()),
+            notify_initial_thread_scan_complete: MockMethod::new_unimplemented(),
+            supports_return_barrier: MockMethod::new_unimplemented(),
+            prepare_for_roots_re_scanning: MockMethod::new_unimplemented(),
+            process_weak_refs: Box::new(MockMethod::<
+                (
+                    &'static mut GCWorker<Self>,
+                    ProcessEdgesWorkTracerContext<SFTProcessEdges<MockVM>>,
+                ),
+                bool,
+            >::new_unimplemented()),
```

Comment:
This method should be similar to the case of `scan_roots_in_mutator_thread`.  It may use the `TracerContext` to create an `ObjectTracer`, and use it to trace several objects.  One way to do mock testing is simply let the closure return a list of `ObjectReference` to be traced.